Improve the upgrade path of Strong Parameters

rafaelfranca opened the pull request:
Some changes were made in this PR to improve the upgrade path of Strong Parameters not inheriting from hash anymore.

### Raise exception when calling `to_h` in an unpermitted `Parameters`
Before we returned either an empty hash or only the always permitted parameters (:controller and :action by default).

The previous behavior was dangerous because in order to get the attributes users usually fallback to use `to_unsafe_h` that could potentially introduce security issues.

The `to_unsafe_h` API is also not good since Parameters is a object that quacks like a `Hash` but not in all cases since `to_h` would return an empty hash and users were forced to check if `to_unsafe_h `is defined or if the instance is a `ActionController::Parameters` in order to work with it. This end up coupling a lot of libraries and parts of the application with something that is from the controller layer.

### Add ActionController::Parameters#to_hash to implicit conversion

Now methods that implicit convert objects to a hash will be able to work without requiring the users to change their implementation.

This method will return a Hash instead of a `HashWithIndefirentAccess` to mimic the same implementation of `HashWithIndefirentAccess#to_hash`.

### Implement ActionController::Parameters#to_query and #to_param

Previously it was raising an error because it may be unsafe to use those methods in a unpermitted parameter. Now we delegate to to_h that already raise an error when the Parameters instance is not permitted.

This also fix a bug when using `#to_query` in a hash that contains a `ActionController::Parameters` instance and was returning the name of the class in the string.

## Backport

The plan is to backport this to 5.1 and part of it to 5.0.

### Backport to 5.0

The plan is to backport the first and the last change fully.

The `#to_hash` change I want to just undeprecate it for now since it is already calling `to_hash` in the internal array and the result includes all the parameters in the case were they are not permitted.

Also implement `respond_to_missing` since we have a `method_missing`.

cc @maartenvg @nicolaslupien @trevorturk

rafaelfranca committed:
Test the correct object

rafaelfranca committed:
Raise exception when calling to_h in a unfiltered Parameters

Before we returned either an empty hash or only the always permitted
parameters (:controller and :action by default).

The previous behavior was dangerous because in order to get the
attributes users usually fallback to use to_unsafe_h that could
potentially introduce security issues.

The to_unsafe_h API is also not good since Parameters is a object that
quacks like a Hash but not in all cases since to_h would return an empty
hash and users were forced to check if to_unsafe_h is defined or if the
instance is a ActionController::Parameters in order to work with it.
This end up coupling a lot of libraries and parts of the application
with something that is from the controller layer.

rafaelfranca committed:
Add ActionController::Parameters#to_hash to implict conversion

Now methods that implicit convert objects to a hash will be able to work
without requiring the users to change their implementation.

This method will return a Hash instead of a HashWithIndefirentAccess
to mimic the same implementation of HashWithIndefirentAccess#to_hash.

rafaelfranca committed:
Add test to make sure that to_unsafe_h don't mutate the target

rafaelfranca committed:
Use the right assetions to better error messages

rafaelfranca committed:
Implement ActionController::Parameters#to_query and #to_param

Previously it was raising an error because it may be unsafe to use those
methods in a unpermitted parameter. Now we delegate to to_h that already
raise an error when the Parameters instance is not permitted.

This also fix a bug when using `#to_query` in a hash that contains a
`ActionController::Parameters` instance and was returning the name of the
class in the string.

kirs commented:
I'm really excited to see `to_h` raising an error if it's not permitted. That would save hours of work when upgrading Shopify to strong parameters.

I have concerns about backporting the last commit to 5.0. The change makes sense in the context of `to_h` raising an error, but that won't be the case for 5.0. Do I miss the point?

rafaelfranca commented:
The plan is to `to_h` raise an error in 5.0 (so backporting the first and last change). An error is better than silently raising making the Parameters empty since now users can know their application will break.

trevorturk commented:
I agree the error is good -- it's much better than silently not working. Good job!

matthewd commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -43,6 +43,18 @@ def initialize(params) # :nodoc:
+  #   # => ActionController::ForbiddenParameters: converting a unpermitted parameters to hash is not allowed
+  class ForbiddenParameters < StandardError
+    def initialize
+      super("converting an unpermitted parameters to hash is not allowed")
```
"unable to convert unpermitted parameters to hash", maybe? "not allowed" sounds like we're being difficult.

matthewd commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -43,6 +43,18 @@ def initialize(params) # :nodoc:
+  #   params = ActionController::Parameters.new(a: "123", b: "456")
+  #   params.to_h
+  #   # => ActionController::ForbiddenParameters: converting a unpermitted parameters to hash is not allowed
+  class ForbiddenParameters < StandardError
```
Forbidden doesn't sound quite right.. how about UnfilteredParameters?

matthewd commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -257,7 +312,7 @@ def to_h
     #   # => {"name"=>"Senjougahara Hitagi", "oddity" => "Heavy stone crab"}
     def to_unsafe_h
-      convert_parameters_to_hashes(@parameters, :to_unsafe_h)
+      permit!.to_h
```
I don't think this should mutate the original params object.. otherwise `params.to_unsafe_h` in one place that's being careful some other way could leave a later `params.to_h` exposed to danger.

rafaelfranca replied:
oh. That is right. I was wondering why we implemented like that and now I know it. We are missing some test coverage about this case. I'll add. Do you have preference between dup the paramters and mutate it or the implementation like it was before?

matthewd replied:
If it's just a simple dup that seems fine.. if it would need to be a deep dup, I think it's probably worth the effort of the more complicated version.

kirs commented:
> The plan is to to_h raise an error in 5.0

That's quite a breaking change to push to stable version. As a developer I would be extremely surprised by this introduced between `5.0.2` and `5.0.3`. This will also make the upgrade to `5.0.3` harder, resulting to lower adoption rate.

At least this should be a config variable, with `raise` disabled by default.

nicolaslupien commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -234,18 +246,61 @@ def ==(other)
+    #   safe_params = params.permit(:name)
+    #   safe_params.to_hash # => {"name"=>"Senjougahara Hitagi"}
+    def to_hash
+      to_h.to_hash
```
This will make our code cleaner, thanks! 👍

matthewd commented:
@kirs that's a good point.. but on the other hand, is anyone going to be knowingly using the current behaviour? It seems a rather long-winded way to spell `{}`.

The low-impact alternative would be to introduce a warning, and optionally offer the raise as a non-default config option. I'm dubious that the exception would cause people problems (as opposed to identifying places they didn't realise were silently eating input), but as 5.0 is fairly mature at this point -- and about to become legacy -- maybe we should go with the more conservative option anyway.

I'm not sure whether this should go through the deprecation infrastructure (gives people more control etc, even though it's not strictly a deprecation), or straight to Kernel#warn.

rafaelfranca commented:
> @kirs that's a good point.. but on the other hand, is anyone going to be knowingly using the current behaviour? It seems a rather long-winded way to spell {}.

Yeah... if people is getting this exception after upgrading to 5.0.3 it means that they had a bug that they didn't anticipated in their code and in my opinion an exception is better than leaving the bug hidden.

We can go in the deprecation path just to be safe with an option to raise on this deprecation.

trevorturk commented:
Yes the deprecation path might be best, like we have `new_rails_defaults` etc. I think raising would have been a better choice to start, but I agree at this point it may be surprising to see new exceptions when you upgrade, so opt-in would be ideal.

rafaelfranca commented:
Fixed all the points for this branch and 5.1. I'll open a new PR with the changes to 5.0

maclover7 commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -234,18 +246,61 @@ def ==(other)
       end
     end
 
+    # Returns a safe <tt>Hash</tt> representation of this parameter
```
the parameters

maclover7 commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -234,18 +246,61 @@ def ==(other)
+    #
+    #   params = ActionController::Parameters.new({
+    #     name: 'David',
+    #     nationality: 'Danish'})
```
maybe have `})` be on a new line, like above?

maclover7 commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -234,18 +246,61 @@ def ==(other)
+    #
+    #   params = ActionController::Parameters.new({
+    #     name: 'David',
+    #     nationality: 'Danish'})
```
same comment -- maybe have `})` be on a new line, like above?

maclover7 reviewed (changes requested):
Left a few comments about the docs :)

kirs commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -43,6 +43,18 @@ def initialize(params) # :nodoc:
+  #   # => ActionController::UnfilteredParameters: unable to convert unpermitted parameters to hash
+  class UnfilteredParameters < StandardError
+    def initialize
+      super("unable to convert unpermitted parameters to hash")
```
It would be friendly to the developer to suggest how they're supposed to fix the error (by using permit)

rafaelfranca replied:
An exception don't need to explain what users have to do. It depends on the context and since we don't know what is the context we can't recommend something. But, it needs to be clear about what it means. Is the exception message and documentation clear about what it means?

matthewd replied:
+1 exceptions should clearly state what's wrong, but leave speculation about how the developer might fix it to someone/something else

rafaelfranca committed:
Improve documentation

We are talking about a list of parameters even so we need to use plural.
Even if we were talking about the instance of the Parameters object we
would have to use the capital and monospaced font.

rafaelfranca commented:
@maclover7 thanks! Done

kaspth commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -228,27 +240,72 @@ def ==(other)
+    #     name: 'David',
+    #     nationality: 'Danish'
+    #   })
+    #   params.to_query('user')
```
Should the new documentation here use double quoted strings and no hash braces on the `new` call to fit our Rubocop rules?

rafaelfranca replied:
Done

kaspth commented on actionpack/test/controller/parameters/parameters_permit_test.rb:
```diff
@@ -377,57 +377,81 @@ def walk_permitted(params)
-    assert_equal({ "controller" => "users", "action" => "create" }, params.to_h)
+      assert_instance_of Hash, params.to_hash
+      assert_not_kind_of ActionController::Parameters, params.to_hash
+      assert_equal({ "crab" => "Senjougahara Hitagi" }, params.to_hash)
```
Wouldn't Ruby implicitly call `to_hash` on the `params` here? Should we test an implicit case?

rafaelfranca replied:
Done

kaspth reviewed (commented):
Like this a lot! 👍

rafaelfranca commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -228,27 +240,72 @@ def ==(other)
       if permitted?
         convert_parameters_to_hashes(@parameters, :to_h)
       else
-        slice(*self.class.always_permitted_parameters).permit!.to_h
```
It is still needed when you permit let say `[:user][:name]` in a controller and the `action_on_unpermitted_parameters` is either `:log` or `:raise`. that way `:controller` and `:action` will not raise.

kaspth replied:
Ah, it's handled in `unpermitted_parameters!`. Never mind 😊

rafaelfranca committed:
Follow the style guide rules in the documetation

rafaelfranca commented on actionpack/lib/action_controller/metal/strong_parameters.rb:
```diff
@@ -43,6 +43,18 @@ def initialize(params) # :nodoc:
   end
 
+  # Raised when a Parameters instance is not marked as permitted and
+  # an operation to transform to hash is called.
```
Done

rafaelfranca committed:
Reuse the Parameters#to_h check in the routing helpers

Since this protection is now in Parameters we can use it instead of
reimplementing again.

kaspth reviewed (approved):
Had one last comment, otherwise 👍

matthewd commented:
There's another stand-alone instance of the existing message in https://github.com/rails/rails/blob/a2ab5ffae97e3523e2d6a8f22222af4df001212f/actionview/lib/action_view/helpers/url_helper.rb#L625-L626

rafaelfranca commented:
Good catch. Fixed

rafaelfranca commented:
Backported to 5-1-stable in a217fd0d13.

PR for 5-0-stable https://github.com/rails/rails/pull/28803.
